### ISSUE TITLE
Interface box template features

### ISSUE THREAD

**Issue by Michiel-s** [AmpersandTarski/Ampersand#1084]
## Problem statement
Currently we have a large set of [BOX templates](https://github.com/AmpersandTarski/prototype/tree/development/templates) that are almost the same but slightly differ in features inside the template. E.g. in the COLS range (tables with where subinterfaces are mapped on columns):
* Box-COLS.html
* Box-COLSNL.html (no labels, i.e. no header row)
* Box-SCOLS.html (sortable table)
* Bob-HCOLS.html (hidden table when no records)
* Box-SHCOLS.html (combi of S and H)

Same holds for ROWS:
* Box-ROWS.html
* Box-ROWSNL.html (no labels)
* Box-HROWS.html (hidden fields when no data)
* Box-HROWSNL.html (combi of H and NL)

Around 95% of these templates are the same. This is hard to maintain and requires new templates for a new combi of features. Also it makes it hard to customize/configure templates for a specific interface. That would introduce another template.

## Feature request
I would like to see the ampersand interface BOX syntax extended to provide configuration to the template mechanism.

E.g.
```adl
INTERFACE "Projects" : V[SESSION*Project] BOX <COLS sort="asc">
    [ "Name"                 : projectName
    , "Projectleader"      : pl
    ]
```
instead of
```adl
INTERFACE "Projects" : V[SESSION*Project] BOX <SCOLS>
    [ "Name"                 : projectName
    , "Projectleader"      : pl
    ]
```
as you may notice, an attribute is added to the BOX <template-name> statement to provide information to the template engine.

## Examples of allowed syntax:
* No attributes. This is current implementation
```
BOX <COLS>
```
* Attribute without value. `sortable` attribute must be available as boolean in template engine, set to true
```
BOX <COLS sortable>
```
* Attribute with string value
```
BOX <COLS sort="asc">
```
* Multiple attributes. All string values!! no other datatypes are needed. The value 5 can also be a string.
```
BOX <COLS sortable sort="asc" limit="5">
```

## Possible difficulties
* The HStringTemplate library we use doesn't allow for string comparison inside the template. So an `$if(sort='asc')$...$endif$` will not be possible. But I think we can live with this limitation
* Is it possible to pass in haskell variables to the template engine that are not known compile time? E.g. all interfaces have the variables `exprIsUni`, `exprIsTot`, `crudR`, `crudC`, etc. But can we make this so that runtime defined variables can be used like `sortable`.

**Comment by hanjoosten** [comment]
If I understand this correctly, you want something like: ([EBNF](http://www.cs.man.ac.uk/~pjj/bnf/bnf.html#EBNF) notation)

```
<BoxTemplate> ::= '<' Identifier keyValPair* '>' 
<keyValPair> ::= identifier '=' quotedString
<quotedString> ::= '"' character+ '"'
```
with `Identifier` and `identifier` represent a single word (no whitespace in them) starting with an uppercase resp. lowercase character. quotedString

Using above syntax, `BoxTemplate` becomes an optional expression following the current "BOX" keyword. 

@Michiel-s , could you confirm this explaination?

**Comment by Michiel-s** [comment]
Yes, but I would like the '=' quotedString part of the keyValPair to be optional.

Not sure if this is the right notation:
`<keyValPair> ::= identifier ('=' quotedString)?`

**Comment by hanjoosten** [comment]
Sure. I might even be able to train the typechecker, to make sure all keys in a single `BoxTemplate` have a different name.

**Comment by hanjoosten** [comment]
I took a shot at this. See the feature branch. Still to do:

- [ ] Typechecker: Make an error when multiple attirbutes have the same name
- [ ] Adapt to the new .json files. (@Michiel-s, could you do that? )

**Comment by Michiel-s** [comment]
Nice. I will check it out.
The json files don’t need this output. Box template attributes are only needed compile time in the template engine. So I will adapt some html templates to test it.

**Comment by Michiel-s** [comment]
I've created a test using the ProjectAdministration case:
https://github.com/Michiel-s/Project-administration/blob/test/ProjectAdministration.adl

To build this prototype, use the following options to set the prototype framework version that includes a Box-TABLE template.
`--prototype-framework-version feature/box-attributes`

There's an interface that looks like this:
```adl
  INTERFACE "Active projects" : V[SESSION*Project];projectActive BOX <TABLE no-header title="List of active projects">
    [ "Projects"      : I[Project]
    , "Name"          : projectName
    , "Description"   : projectDescription
    , "Projectleider" : pl
    , "@"             : pl[Project*Person] <PersonEmail>
    , "Start"         : projectStartDate
    , "Active"        : projectActive
    ]
```

**Comment by hanjoosten** [comment]
The TABLE template currently does not parse. @Michiel-s , if you fix the template, it should do some amazing stuff. Give it a try and play around with templates. Probably there are some more things to wish for. We'll see. Let me know how things work out.

**Comment by hanjoosten** [comment]
Some comments on what has changed so far:

- The ADL-parser has been changed to accomodate the BOX specification, as requested.
- The generated interfaces.json file reflects those changes.
- The handling of the template files is now different:
   1. Rendering starts by filling in the named attributes in the template that match any key in the BOX specification. An attribute that has a key without a string will be given the value `true`.
   2. Rendering continues by filling in specific remaining named attributes. These are the attributes with known names like `ampersandVersionStr` or `crudC`. The values will be filled in in the current way, as you would expect. Notice that if you specify such an attribute in the BOX specification, it would be rendered in the first step. So it is possible to override these values. 
   3. The remaining named attributes will be given the value `false`. There is no way to distinquish based on type, because in the template every attribute is of type String. We could however do something based on some naming convention. @Michiel-s, you might have ideas about that.

**Comment by Michiel-s** [comment]
@hanjoosten: please also allow `-` character in box template attribute keys.
E.g. `BOX <TABLE no-header title="List of active projects">`

**Comment by hanjoosten** [comment]
This is a tricky request. I do think that using [kebab-case](https://stackoverflow.com/questions/11273282/whats-the-name-for-hyphen-separated-case) for names is nice. However, in Ampersand we use snake_case. This has historically grown, because we looked at what is common with names in database environments. 
The Ampersand lexer/parser can see an underscore character as part of a name. However,  `-` is a token in itself. TLDR: This would involve some deep changes in the parser, probably (to do it right) also in the lexer.

This said, I don't like to use an argument that simply says that it is a lot of work and/or it is difficult as a single argument not to do stuff, but I do like to keep things simple.  Using kebab-case for these names in ADL and convert them to something else because hstringtemplate might not be able to cope with it might in the end be confusing for the user too. 
I wonder if all of this is worth the trouble.

**Comment by Michiel-s** [comment]
Ok. let's don't do that. We'll use camelCase for BOX header attributes instead.

**Comment by Michiel-s** [comment]
Implemented by #1087
